widgetsnbextension/src/manager.js

```diff
@@ -421,12 +430,81 @@ WidgetManager.prototype._updateCellSnapshots = function(cell, index) {
         // Create a mime bundle.
         const bundle = {};
         bundle[imageMimetype] = imageData;
+        bundle['text/html'] = widgets.generateEmbedScript(
+            widgetState,
+            'Could not render widget, embed-manager missing from page context.'
+        );
         widgetSubarea.widgetSnapshot = bundle;
     }).then(function() {
         return that.progressModal.setValue(++that._progress/cells.length);
     });
 };
 
+WidgetManager.prototype._getCellWidgetStates = function(cell, index) {
+    var that = this;
+    var modelIds = Object.keys(this._models);
+    return Promise.all(modelIds.map(function(modelId) {
+        return that._models[modelId].then(function(model) {
+            if (model.options.cell_index === index) {
+                return that
+                    ._traverseWidgetTree(model)
+                    .map(function(widget) { return widget.get_state(true); });
+            } else {
+                return [null];
+            }
+        });
+    })).then(function(states) {
+        return states
+            .reduce(function(a, b) { return a.concat(b); }, [])
+            .filter(function(state) { return state !== null; });
+    });
+};
+
+WidgetManager.prototype._traverseWidgetTree = function(parentWidget, cache) {
+    // Setup a cache if it doesn't already exist
+    if (!cache) {
+        cache = {};
+    }
+
+    // Don't continue if this widget has already been traversed
+    if (parentWidget.id in cache) {
+        return [];
+    }
+
+    // Remember that this widget has been traversed
+    cache[parentWidget.id] = true;
+
+    // Traverse the parent widget's state for child widgets
+    const state = parentWidget.get_state(true);
+    const subWidgets = this._traverseWidgetState(state);
+    const that = this;
+    return _.flatten(subWidgets.map(function(subWidget) {
+        return that._traverseWidgetTree(subWidget, cache);
+    })).concat([ parentWidget ]);
+};
+
+WidgetManager.prototype._traverseWidgetState = function(state) {
+    var that = this;
+    if (state instanceof widgets.Widget) {
+        return [state];
+    } else if (_.isArray(state)) {
+        return _.flatten(state.map(this._traverseWidgetState.bind(this)));
+    } else if (state instanceof Object) {
+        var states = [];
+        _.each(state, function(value, key) {
+            states.push(that._traverseWidgetState(value));
+        });
+        return _.flatten(states);
+    } else {
+        return [];
+    }
+};
```

Comment:
At the moment, we only reference peer widget models that are listened to as as child widgets, which would ensure that this works, but it is not clear that it would always be the case.